pkg/service/service_test.go

```diff
@@ -1484,6 +1509,71 @@ func (m *ManagerTestSuite) TestUpsertServiceWithZeroWeightBackends(c *C) {
 	c.Assert(m.lbmap.DummyMaglevTable[uint16(id1)], Equals, 1)
 }
 
+func (m *ManagerTestSuite) TestUpdateBackendsStateWithBackendSharedAcrossServices(c *C) {
+	option.Config.NodePortAlg = option.NodePortAlgMaglev
+	be := append(backends1, backends4...)
+	backends := make([]*lb.Backend, 0, len(be))
+	for _, b := range be {
+		backends = append(backends, b.DeepCopy())
+	}
+	backends[0].State = lb.BackendStateActive
+	backends[1].State = lb.BackendStateActive
+	backends[2].State = lb.BackendStateMaintenance
+	hash0 := backends[0].L3n4Addr.Hash()
+	hash1 := backends[1].L3n4Addr.Hash()
+	hash2 := backends[2].L3n4Addr.Hash()
+
+	p := &lb.SVC{
+		Frontend:                  frontend1,
+		Backends:                  backends,
+		Type:                      lb.SVCTypeNodePort,
+		ExtTrafficPolicy:          lb.SVCTrafficPolicyCluster,
+		IntTrafficPolicy:          lb.SVCTrafficPolicyCluster,
+		SessionAffinity:           true,
+		SessionAffinityTimeoutSec: 100,
+		Name: lb.ServiceName{
+			Name:      "svc1",
+			Namespace: "ns1",
+		},
+	}
+	r := &lb.SVC{
+		Frontend:                  frontend2,
+		Backends:                  backends,
+		Type:                      lb.SVCTypeNodePort,
+		ExtTrafficPolicy:          lb.SVCTrafficPolicyCluster,
+		IntTrafficPolicy:          lb.SVCTrafficPolicyCluster,
+		SessionAffinity:           true,
+		SessionAffinityTimeoutSec: 100,
+		Name: lb.ServiceName{
+			Name:      "svc2",
+			Namespace: "ns1",
+		},
+	}
+	svcHash2 := r.Frontend.Hash()
+
+	_, id1, err := m.svc.UpsertService(p)
+	c.Assert(err, IsNil)
+	_, id1, err = m.svc.UpsertService(r)
+	c.Assert(err, IsNil)
+	_, id1, err = m.svc.UpsertService(r)
```

Comment:
I'm trying to understand the minimal set of changes for the unit test case. Does the `UpsertService` for the 2nd service required?
Or it it enough to have a single `UpsertService(s)` call followed by the `UpdateBackendsState` and then `UpsertService(s)` call?

It would help if you can call out the set of commands that you ran when you originally discovered this issue.

Reply:
@aditighag 

> It would help if you can call out the set of commands that you ran

yeah, I get it.... unfortunately, our use case is a program that synces our balancing using the ciliium API, so the way we discovered this bug was just "after a while of random changes in the cluster, some services seem to ignore any new changes to them" so I actually digged the bug right out of code, not by trying out different CLI/API calls manually.

> I'm trying to understand the minimal set of changes

I understand, sorry, it's a little over-complicated thing with the pointers. So the minimal number of changes for the pointers to break is `upsert(r)` and `upsert(r)` again. BUT if you only did that, the only way I would be able to demonstrate the bug is that at this point (after the 2 upserts) the only thing that is seemingly wrong would be the pointers not pointing to the same object ( see the PR description above: `a service is upserted with backend already existing in 3.`)

Unfortunately, if we only leave the 2 `upsert(r)` the existing asserts would still be okay, because (as stated in the bullet point I mentioned) `1. and 3. stil point to the same objects` therefore the `UpdateBackendsState` would "accidentally" update the object to the correct state [here](https://github.com/cilium/cilium/blob/ec8d3ced8e2d6300e06b6f9ddc83d00f3cad6157/pkg/service/service.go#L818)

TLDR; from code perspective `upsert(r)` -> `upsert(r)` is what produces the bug but from user-perspective `upsert(p)` -> `upsert(r)`  -> `upsert(r)` needs to be performed so that `UpdateBackendsState` called **visibly** breaks the state for the user.

Try to reach me on slack if still in doubt.

Reply:
Ok, thanks for the additional context. Did you confirm that the unit test fails without your [fix](https://github.com/cilium/cilium/pull/24474/commits/96b822f0fb0775f3cf8fc09d9b8fb3475478151d)?

I'm being cautious here, because any bugs in this code could have severe effects on connectivity if the backends are not correctly plumbed in the datapath so appreciate your help.

Reply:
yep, just try dropping the added unit test as-is in the `main` branch, fails like a glove.